File: packages/api/src/window.ts
import { VantailError } from "./error.js";
import { invoke, listen, windowLabel } from "./transport.js";
import type { TitleBarMetrics, VantailBridge } from "./protocol.js";

export interface Size {
  width: number;
  height: number;
}

export interface Position {
  x: number;
  y: number;
}

/** Everything a new window can be given, matching `window` in the config. */
export interface WindowOptions {
  /** Path within the application, e.g. `settings.html` or `/#/settings`. */
  url?: string;
  /**
   * Resolve only once the new window's page has loaded. Default `true`.
   *
   * A window exists before its document does, so without this a message sent
   * straight after `createWindow` would arrive at a webview with nothing
   * listening yet, and vanish.
   */
  waitForReady?: boolean;
  /** How long to wait for that. Default 15000. */
  readyTimeoutMs?: number;
  title?: string;
  width?: number;
  height?: number;
  minWidth?: number;
  minHeight?: number;
  maxWidth?: number;
  maxHeight?: number;
  x?: number;
  y?: number;
  resizable?: boolean;
  maximized?: boolean;
  fullscreen?: boolean;
  decorations?: boolean;
  /**
   * Whether the title bar is a bar, or space your application draws in.
   *
   * `hidden` is what every editor and browser does: no title bar, the page
   * running to the top edge of the window, and a toolbar of your own where
   * the bar would have been.
   *
   * On macOS the traffic lights stay - they are the system's, and an
   * application drawing its own gets them subtly wrong. Windows and Linux
   * have no way to keep the buttons without the bar, so `hidden` there is an
   * undecorated window and your toolbar has to include close and minimise.
   *
   * A window with no title bar has nothing to drag it by. Give your toolbar
   * `appWindow.startDragging()` on `pointerdown`.
   */
  titleBarStyle?: TitleBarStyle;
  /**
   * Nudge the traffic lights, for a toolbar taller than the bar it replaced.
   * Logical pixels from the top left. macOS only.
   */
  trafficLightPosition?: { x: number; y: number };
  /**
   * How tall the bar your application draws should be.
   *
   * Defaults to the height of the platform's own, which is what makes a
   * custom bar read as a title bar rather than as a div. Set it larger for a
   * browser-style toolbar and the traffic lights are re-centred in it - the
   * part that is easy to get wrong by hand, and obvious the moment it is.
   */
  titleBarHeight?: number;
  /**
   * Whether the platform draws the window buttons, or you do.
   *
   * macOS keeps its traffic lights when the title bar is hidden, and they are
   * a fixed size. `hidden` takes them away so you can draw your own - bigger,
   * or in your own style - the way you already have to on the platforms that
   * keep nothing.
   *
   * With them hidden, `insetLeft` is `0`: the same signal those platforms
   * already give, so code that draws its own controls when nothing is
   * reserved needs no new branch.
   */
  titleBarButtons?: TitleBarButtons;
  transparent?: boolean;
  alwaysOnTop?: boolean;
  center?: boolean;
  visible?: boolean;
  closeBehavior?: CloseBehavior;
}

/**
 * What a window's own close button does.
 *
 * - `close` - destroy it. With no other windows open and
 *   `quitOnLastWindowClosed`, that ends the application.
 * - `hide` - hide it instead. The webview keeps running, so timers keep
 *   firing and sockets stay connected. This is what a tray application wants.
 * - `ask` - do nothing, and let `onCloseRequested` decide.
 *
 * There is no `preventDefault()` here on purpose: it would mean the runtime
 * waiting on an answer from JavaScript before it could close a window, and an
 * application that never answered would have a window nobody could shut.
 * Declaring the behaviour is the same thing without that failure mode.
 */
export type CloseBehavior = "close" | "hide" | "ask";

/**
 * `default` keeps the platform's title bar. `hidden` removes it and lets the
 * page run to the top edge of the window - see `titleBarStyle`.
 */
export type TitleBarStyle = "default" | "hidden";

/** Who draws close, minimise and zoom. */
export type TitleBarButtons = "system" | "hidden";

/**
 * The room a hidden title bar left behind, without awaiting.
 *
 * `undefined` outside a Vantail window. All zeroes when this window has an
 * ordinary title bar, since there is then nothing to leave room for.
 *
 * The same numbers are already on `:root` as CSS variables before the page
 * lays out, so reach for those first:
 *
 * ```css
 * .toolbar {
 *   height: var(--vantail-titlebar-height);
 *   padding-left: var(--vantail-titlebar-inset-left);
 *   padding-right: var(--vantail-titlebar-inset-right);
 * }
 * ```
 *
 * This is for the cases CSS cannot reach - laying out a canvas, or deciding
 * whether to draw your own window controls:
 *
 * ```ts
 * const { insetLeft } = titleBarMetrics() ?? { insetLeft: 0 };
 * if (insetLeft === 0) drawOwnWindowControls();
 * ```
 */
export function titleBarMetrics(): TitleBarMetrics | undefined {
  return (globalThis as { __VANTAIL__?: VantailBridge }).__VANTAIL__?.titleBar;
}

/**
 * A handle to one window.
 *
 * `appWindow` is the window your code is running in. `getWindow(label)` is any
 * other. The two are the same object shape, so code that adjusts a window does
 * not care which one it was handed.
 */
export interface WindowHandle {
  /** `undefined` on `appWindow`, which always means "whichever window I am". */
  readonly label: string | undefined;

  setTitle(title: string): Promise<null>;
  title(): Promise<string>;

  setSize(width: number, height: number): Promise<null>;
  size(): Promise<Size>;
  setPosition(x: number, y: number): Promise<null>;
  position(): Promise<Position>;
  center(): Promise<null>;

  minimize(): Promise<null>;
  unminimize(): Promise<null>;
  /**
   * Stop the user shrinking the window below this. Pass `null` to remove it.
   *
   * Logical pixels, like every other window measurement.
   */
  setMinSize(width: number | null, height: number | null): Promise<null>;
  /** The other end. `null` removes it. */
  setMaxSize(width: number | null, height: number | null): Promise<null>;
  /**
   * Keep the window out of the taskbar. Windows and Linux only - macOS has no
   * taskbar, and answers `UNSUPPORTED`.
   */
  setSkipTaskbar(value: boolean): Promise<null>;

  maximize(): Promise<null>;
  unmaximize(): Promise<null>;
  toggleMaximize(): Promise<boolean>;
  isMaximized(): Promise<boolean>;

  setFullscreen(value: boolean): Promise<null>;
  isFullscreen(): Promise<boolean>;
  setResizable(value: boolean): Promise<null>;
  setAlwaysOnTop(value: boolean): Promise<null>;

  /**
   * Drag the window from wherever the pointer is.
   *
   * What a title bar would have done for you. Call it on `pointerdown` in
   * your own toolbar and the platform takes the drag from there:
   *
   * ```ts
   * toolbar.addEventListener("pointerdown", (event) => {
   *   // Let buttons and inputs be clicked rather than dragged.
   *   if ((event.target as Element).closest("button, input, a")) return;
   *   if (event.buttons === 1) void appWindow.startDragging();
   * });
   * ```
   *
   * `-webkit-app-region: drag` is a Chromium extension and does nothing in a
   * WKWebView, which is why this is a call rather than a CSS property.
   */
  startDragging(): Promise<null>;

  /**
   * Switch between an ordinary title bar and one your application draws in,
   * on a window that is already open.
   *
   * Answers with the room the new arrangement leaves. The CSS variables and
   * {@link titleBarMetrics} are updated before it resolves, so a toolbar
   * sized from either resizes with the change.
   *
   * ```ts
   * const { height } = await appWindow.setTitleBarStyle("hidden");
   * ```
   *
   * On macOS the switch is seamless. Everywhere else the only lever is the
   * window frame, so this adds and removes decorations - and switching back
   * restores what the config asked for rather than assuming a frame.
   */
  setTitleBarStyle(style: TitleBarStyle): Promise<TitleBarMetrics>;

  /** Which one this window is using now. */
  titleBarStyle(): Promise<TitleBarStyle>;

  /**
   * Hand the window buttons to your application, or take them back.
   *
   * ```ts
   * await appWindow.setTitleBarButtons("hidden"); // draw your own
   * ```
   *
   * Answers with the metrics, whose `insetLeft` is then `0`.
   */
  setTitleBarButtons(buttons: TitleBarButtons): Promise<TitleBarMetrics>;

  /** Which of the two this window is using. */
  titleBarButtons(): Promise<TitleBarButtons>;

  /**
   * Ask for a bar of a particular height; `null` for the platform's own.
   *
   * ```ts
   * await appWindow.setTitleBarHeight(48); // a browser-style toolbar
   * await appWindow.setTitleBarHeight(null); // back to the platform's
   * ```
   *
   * The traffic lights move to the middle of the new height, and the CSS
   * variables follow, so a toolbar sized from them resizes with it.
   */
  setTitleBarHeight(height: number | null): Promise<TitleBarMetrics>;

  /**
   * Move the traffic lights, for a toolbar taller than the bar they came
   * from. Logical pixels from the top left.
   *
   * macOS only; `UNSUPPORTED` everywhere else, since nowhere else has them.
   */
  setTrafficLightPosition(x: number, y: number): Promise<TitleBarMetrics>;

  /**
   * Put them back in the middle of the bar, which is where they sit unless
   * you have moved them.
   */
  centerTrafficLights(): Promise<TitleBarMetrics>;

  setCloseBehavior(behavior: CloseBehavior): Promise<null>;
  closeBehavior(): Promise<CloseBehavior>;

  show(): Promise<null>;
  hide(): Promise<null>;
  isVisible(): Promise<boolean>;
  focus(): Promise<null>;
  close(): Promise<boolean>;
  exists(): Promise<boolean>;

  openDevtools(): Promise<null>;

  /**
   * Window events reach only the window they happened to, so these fire for
   * `appWindow` and stay quiet on a handle to somebody else's window.
   */
  onResized(handler: (size: Size) => void): () => void;
  onMoved(handler: (position: Position) => void): () => void;
  onFocusChanged(handler: (state: { focused: boolean }) => void): () => void;
  /**
   * The close button was pressed. `outcome` says what the runtime did:
   * `hidden` under `hide`, or `ignored` under `ask` - in which case closing
   * or hiding the window is now your job.
   *
   * Not emitted under `close`, where the window is already gone.
   */
  onCloseRequested(
    handler: (event: { outcome: "hidden" | "ignored" }) => void,
  ): () => void;
}

function handle(label?: string): WindowHandle {
  // Omitting `label` entirely lets the runtime answer for the calling window,
  // which is what makes one implementation serve both cases.
  const target = label === undefined ? {} : { label };
  const call = <T>(method: string, params: object = {}) =>
    invoke<T>(method, { ...target, ...params });

  const scoped = <T extends { label?: string }>(
    event: string,
    handler: (payload: T) => void,
  ) => {
    // `appWindow` has no label of its own, but it still means *this* window -
    // so fall back to the one the runtime injected rather than taking
    // everything that happens to arrive.
    const mine = label ?? windowLabel();
    return listen<T>(event, (payload) => {
      if (
        mine !== undefined &&
        payload.label !== undefined &&
        payload.label !== mine
      )
        return;
      handler(payload);
    });
  };

  return {
    label,

    setTitle: (title) => call<null>("window.setTitle", { title }),
    title: () => call<string>("window.title"),

    setSize: (width, height) => call<null>("window.setSize", { width, height }),
    size: () => call<Size>("window.size"),
    setPosition: (x, y) => call<null>("window.setPosition", { x, y }),
    position: () => call<Position>("window.position"),
    center: () => call<null>("window.center"),

    minimize: () => call<null>("window.minimize"),
    unminimize: () => call<null>("window.unminimize"),
    setMinSize: (width: number | null, height: number | null) =>
      call<null>("window.setMinSize", { width, height }),
    setMaxSize: (width: number | null, height: number | null) =>
      call<null>("window.setMaxSize", { width, height }),
    setSkipTaskbar: (value: boolean) =>
      call<null>("window.setSkipTaskbar", { value }),

    maximize: () => call<null>("window.maximize"),
    unmaximize: () => call<null>("window.unmaximize"),
    toggleMaximize: () => call<boolean>("window.toggleMaximize"),
    isMaximized: () => call<boolean>("window.isMaximized"),

    setFullscreen: (value) => call<null>("window.setFullscreen", { value }),
    isFullscreen: () => call<boolean>("window.isFullscreen"),
    setResizable: (value) => call<null>("window.setResizable", { value }),
    setAlwaysOnTop: (value) => call<null>("window.setAlwaysOnTop", { value }),

    setTitleBarStyle: (style) =>
      call<TitleBarMetrics>("window.setTitleBarStyle", { style }),
    titleBarStyle: () => call<TitleBarStyle>("window.titleBarStyle"),
    setTitleBarButtons: (buttons) =>
      call<TitleBarMetrics>("window.setTitleBarButtons", { buttons }),
    titleBarButtons: () => call<TitleBarButtons>("window.titleBarButtons"),
    setTitleBarHeight: (height) =>
      call<TitleBarMetrics>("window.setTitleBarHeight", { height }),

    startDragging: () => call<null>("window.startDragging"),
    setTrafficLightPosition: (x, y) =>
      call<TitleBarMetrics>("window.setTrafficLightPosition", { x, y }),
    centerTrafficLights: () =>
      call<TitleBarMetrics>("window.centerTrafficLights"),

    setCloseBehavior: (behavior) =>
      call<null>("window.setCloseBehavior", { behavior }),
    closeBehavior: () => call<CloseBehavior>("window.closeBehavior"),

    show: () => call<null>("window.show"),
    hide: () => call<null>("window.hide"),
    isVisible: () => call<boolean>("window.isVisible"),
    focus: () => call<null>("window.focus"),
    close: () => call<boolean>("window.close"),
    exists: () => call<boolean>("window.exists"),

    openDevtools: () => call<null>("window.openDevtools"),

    onResized: (fn) => scoped<Size & { label?: string }>("window.resized", fn),
    onMoved: (fn) => scoped<Position & { label?: string }>("window.moved", fn),
    onFocusChanged: (fn) =>
      scoped<{ focused: boolean; label?: string }>("window.focus", fn),
    onCloseRequested: (fn) =>
      scoped<{ outcome: "hidden" | "ignored"; label?: string }>(
        "window.closeRequested",
        fn,
      ),
  };
}

/**
 * The window this code is running in.
 *
 * Note this shadows the global `window` when imported under that name. Import
 * it as `appWindow` if you need both.
 */
export const appWindow: WindowHandle = handle();

/** A handle to another window. Calls on a window that is gone reject. */
export function getWindow(label: string): WindowHandle {
  return handle(label);
}

/**
 * Open a new window and return a handle to it.
 *
 * Labels are how every other call names this window, so they have to be
 * unique - creating one that already exists rejects with `ALREADY_EXISTS`.
 */
export async function createWindow(
  label: string,
  options: WindowOptions = {},
): Promise<WindowHandle> {
  const {
    url,
    waitForReady = true,
    readyTimeoutMs = 15_000,
    ...window
  } = options;

  // Subscribed before the window is asked for, so a page that loads instantly
  // cannot report ready before anyone is listening.
  const ready = waitForReady ? waitFor(label, readyTimeoutMs) : undefined;

  try {
    await invoke<string>("window.create", {
      label,
      ...(url === undefined ? {} : { url }),
      window,
    });
  } catch (error) {
    ready?.cancel();
    throw error;
  }

  await ready?.promise;
  return handle(label);
}

function waitFor(
  label: string,
  timeoutMs: number,
): { promise: Promise<void>; cancel(): void } {
  let settle: (() => void) | undefined;
  let fail: ((error: Error) => void) | undefined;

  const stop = listen<{ label: string }>("window.ready", (event) => {
    if (event.label === label) settle?.();
  });

  const timer = setTimeout(() => {
    fail?.(
      new VantailError(
        "INTERNAL",
        `The window "${label}" opened but its page did not finish loading within ${timeoutMs}ms. ` +
          `Check the URL it was given, or pass { waitForReady: false }.`,
      ),
    );
  }, timeoutMs);

  const promise = new Promise<void>((resolve, reject) => {
    settle = resolve;
    fail = reject;
  }).finally(() => {
    clearTimeout(timer);
    stop();
  });

  return {
    promise,
    cancel: () => {
      clearTimeout(timer);
      stop();
      // Nothing is awaiting it, but an unhandled rejection would still be
      // reported if the timer had already fired.
      settle?.();
    },
  };
}

/** Fires when a window's page has finished loading. */
export function onWindowReady(
  handler: (event: { label: string }) => void,
): () => void {
  return listen<{ label: string }>("window.ready", handler);
}

/** Labels of every open window, in the order they were opened. */
export function listWindows(): Promise<string[]> {
  return invoke<string[]>("window.list");
}

/** The label of the calling window, without awaiting. */
export function currentWindow(): string | undefined {
  return windowLabel();
}

export function onWindowCreated(
  handler: (event: { label: string }) => void,
): () => void {
  return listen<{ label: string }>("window.created", handler);
}

export function onWindowClosed(
  handler: (event: { label: string }) => void,
): () => void {
  return listen<{ label: string }>("window.closed", handler);
}

File: runtime/src/api/window.rs
//! `window.*` - the application's windows.
//!
//! Every method acts on the window that called it unless the call names a
//! different one with `label`. That keeps the common case - an app with one
//! window, or a window adjusting itself - free of ceremony, while still
//! letting any window drive any other.

use serde::Deserialize;
use serde_json::{json, Value};
use tao::dpi::{LogicalPosition, LogicalSize};
use tao::window::Fullscreen;

use crate::config::{CloseBehavior, WindowConfig};
use crate::error::{ApiError, ApiResult};
use crate::ipc::{Event, Outgoing, Request};
use crate::state::MainCtx;
use crate::windows;

#[derive(Deserialize, Default)]
#[serde(default)]
struct Target {
    label: Option<String>,
}

#[derive(Deserialize)]
struct Title {
    title: String,
}

#[derive(Deserialize)]
struct Size {
    width: f64,
    height: f64,
}

#[derive(Deserialize)]
struct Position {
    x: f64,
    y: f64,
}

#[derive(Deserialize)]
struct Flag {
    value: bool,
}

#[derive(Deserialize)]
struct Style {
    style: crate::config::TitleBarStyle,
}

#[derive(Deserialize)]
struct Buttons {
    buttons: crate::config::TitleBarButtons,
}

#[derive(Deserialize)]
struct Height {
    /// `None` means the platform's own.
    #[serde(default)]
    height: Option<f64>,
}

#[derive(Deserialize)]
struct Behaviour {
    behavior: CloseBehavior,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct CreateParams {
    label: String,
    /// Path within the application, e.g. `settings.html` or `/#/settings`.
    #[serde(default)]
    url: Option<String>,
    #[serde(default)]
    window: WindowConfig,
}

#[derive(Debug, serde::Deserialize)]
struct Limit {
    width: Option<f64>,
    height: Option<f64>,
}

#[derive(Debug, serde::Deserialize)]
struct Toggle {
    value: bool,
}

/// A limit is only a limit if both dimensions are given: tao takes one size,
/// not a width and a height that can be constrained separately.
fn size_limit(width: Option<f64>, height: Option<f64>) -> Option<LogicalSize<f64>> {
    match (width, height) {
        (Some(width), Some(height)) => Some(LogicalSize::new(width, height)),
        _ => None,
    }
}

/// Maximise or restore, keeping size limits out of the way while maximised.
///
/// A maximised window has to be exactly the size the window system chose. On
/// Wayland that is not a preference: the compositor sends a configure, and a
/// surface that then commits a different geometry is killed for violating
/// xdg-shell -
///
///   xdg_surface geometry (468 x 328) does not match the configured
///   maximized state (1400 x 968)
///
/// which takes the whole application with it. A maximum smaller than the
/// screen does not constrain maximising there anyway, so dropping the limits
/// while maximised costs nothing and removes the way to be killed. They go
/// back on when the window is restored.
fn set_maximized(entry: &mut crate::windows::WindowEntry, maximized: bool) {
    if maximized {
        entry.window.set_min_inner_size(NO_LIMIT);
        entry.window.set_max_inner_size(NO_LIMIT);
        entry.window.set_maximized(true);
        return;
    }

    entry.window.set_maximized(false);
    entry.window.set_min_inner_size(entry.min_size);
    entry.window.set_max_inner_size(entry.max_size);
}

/// Spelled out because `None` alone leaves the size type ambiguous.
const NO_LIMIT: Option<LogicalSize<f64>> = None;

/// macOS has no taskbar, so there is nothing to skip.
#[allow(clippy::needless_return)]
fn skip_taskbar(window: &tao::window::Window, skip: bool) -> ApiResult {
    #[cfg(any(target_os = "windows", target_os = "linux"))]
    {
        #[cfg(target_os = "linux")]
        use tao::platform::unix::WindowExtUnix;
        #[cfg(target_os = "windows")]
        use tao::platform::windows::WindowExtWindows;

        return window
            .set_skip_taskbar(skip)
            .map(|_| Value::Null)
            .map_err(|e| ApiError::internal(format!("Could not change the taskbar: {e}")));
    }

    #[cfg(not(any(target_os = "windows", target_os = "linux")))]
    {
        let _ = (window, skip);
        Err(ApiError::unsupported(
            "macOS has no taskbar to skip. Hide the window instead, or use a tray icon.",
        ))
    }
}

pub fn dispatch(ctx: &mut MainCtx<'_>, method: &str, params: Value) -> ApiResult {
    ctx.rt
        .permissions
        .require(ctx.rt.permissions.window, method)?;

    // Calls about the set of windows rather than one window.
    match method {
        "window.create" => return create(ctx, params),
        "window.list" => return Ok(json!(ctx.windows.labels())),
        "window.current" => return Ok(json!(ctx.source)),
        "window.exists" => {
            let target: Target = serde_json::from_value(params).unwrap_or_default();
            let label = target.label.unwrap_or_else(|| ctx.source.to_string());
            return Ok(json!(ctx.windows.get(&label).is_some()));
        }
        "window.close" => return close(ctx, params),
        _ => {}
    }

    let label = target_label(ctx, &params);
    let entry = ctx.windows.require_mut(&label)?;
    let window = &entry.window;

    match method {
        "window.setTitle" => {
            let Title { title } = Request::params(method, params)?;
            window.set_title(&title);
            Ok(Value::Null)
        }
        "window.title" => Ok(json!(window.title())),

        "window.minimize" => {
            window.set_minimized(true);
            Ok(Value::Null)
        }
        "window.unminimize" => {
            window.set_minimized(false);
            Ok(Value::Null)
        }
        "window.maximize" => {
            set_maximized(entry, true);
            Ok(Value::Null)
        }
        "window.unmaximize" => {
            set_maximized(entry, false);
            Ok(Value::Null)
        }
        "window.toggleMaximize" => {
            let now = !entry.window.is_maximized();
            set_maximized(entry, now);
            Ok(json!(now))
        }
        "window.isMaximized" => Ok(json!(window.is_maximized())),

        "window.setSize" => {
            let Size { width, height } = Request::params(method, params)?;
            window.set_inner_size(LogicalSize::new(width, height));
            // Record it now so a `size()` immediately after `setSize()` is
            // right, rather than waiting for the resize event to catch up.
            entry.size = LogicalSize::new(width, height);
            Ok(Value::Null)
        }
        "window.size" => Ok(json!({ "width": entry.size.width, "height": entry.size.height })),

        // `null` for either dimension removes that limit.
        "window.setMinSize" => {
            let Limit { width, height } = Request::params(method, params)?;
            let limit = size_limit(width, height);
            entry.min_size = limit;
            if !entry.window.is_maximized() {
                entry.window.set_min_inner_size(limit);
            }
            Ok(Value::Null)
        }
        "window.setMaxSize" => {
            let Limit { width, height } = Request::params(method, params)?;
            let limit = size_limit(width, height);
            entry.max_size = limit;
            if !entry.window.is_maximized() {
                entry.window.set_max_inner_size(limit);
            }
            Ok(Value::Null)
        }

        "window.setSkipTaskbar" => {
            let Toggle { value } = Request::params(method, params)?;
            skip_taskbar(window, value)
        }

        "window.setPosition" => {
            let Position { x, y } = Request::params(method, params)?;
            window.set_outer_position(LogicalPosition::new(x, y));
            Ok(Value::Null)
        }
        "window.position" => {
            let position = window
                .outer_position()
                .map_err(|e| ApiError::unsupported(format!("Window position is unavailable: {e}")))?
                .to_logical::<f64>(window.scale_factor());
            Ok(json!({ "x": position.x, "y": position.y }))
        }

        "window.center" => {
            windows::center(window);
            Ok(Value::Null)
        }

        "window.setFullscreen" => {
            let Flag { value } = Request::params(method, params)?;
            window.set_fullscreen(value.then(|| Fullscreen::Borderless(None)));
            Ok(Value::Null)
        }
        "window.isFullscreen" => Ok(json!(window.fullscreen().is_some())),

        "window.setResizable" => {
            let Flag { value } = Request::params(method, params)?;
            window.set_resizable(value);
            Ok(Value::Null)
        }
        "window.setCloseBehavior" => {
            let Behaviour { behavior } = Request::params(method, params)?;
            entry.close_behavior = behavior;
            Ok(Value::Null)
        }
        "window.closeBehavior" => Ok(json!(match entry.close_behavior {
            CloseBehavior::Close => "close",
            CloseBehavior::Hide => "hide",
            CloseBehavior::Ask => "ask",
        })),

        "window.setAlwaysOnTop" => {
            let Flag { value } = Request::params(method, params)?;
            window.set_always_on_top(value);
            Ok(Value::Null)
        }

        "window.show" => {
            window.set_visible(true);
            Ok(Value::Null)
        }
        "window.hide" => {
            window.set_visible(false);
            Ok(Value::Null)
        }
        "window.isVisible" => Ok(json!(window.is_visible())),
        "window.focus" => {
            window.set_focus();
            Ok(Value::Null)
        }

        // A window with no title bar has nothing to drag it by, so the
        // application's own toolbar has to do it. `-webkit-app-region: drag`
        // is a Chromium extension and does nothing in a WKWebView, so this is
        // the portable way: call it from `pointerdown` and the platform takes
        // over the drag from there.
        "window.startDragging" => {
            window
                .drag_window()
                .map_err(|e| ApiError::unsupported(format!("Could not drag the window: {e}")))?;
            Ok(Value::Null)
        }

        // Switching between an ordinary title bar and one the application
        // draws in, on a window that is already open. Answers with the room
        // the new arrangement leaves, since that is what a caller needs next.
        "window.setTitleBarStyle" => {
            let Style { style } = Request::params(method, params)?;
            let label = entry.label.clone();
            let entry = ctx.windows.require_mut(&label)?;
            Ok(json!(entry.set_title_bar_style(style)))
        }
        // A taller bar than the platform's, with the lights re-centred in it.
        // `null` puts it back to whatever the platform uses.
        "window.setTitleBarHeight" => {
            let Height { height } = Request::params(method, params)?;
            let label = entry.label.clone();
            let entry = ctx.windows.require_mut(&label)?;
            Ok(json!(entry.set_title_bar_height(height)))
        }

        "window.titleBarStyle" => Ok(json!(match entry.title_bar_style {
            crate::config::TitleBarStyle::Hidden => "hidden",
            crate::config::TitleBarStyle::Default => "default",
        })),

        // Where the traffic lights sit, for a toolbar taller than the bar it
        // replaced. macOS only, because nowhere else has them.
        "window.setTrafficLightPosition" => {
            if cfg!(not(target_os = "macos")) {
                return Err(ApiError::unsupported(
                    "Only macOS has traffic lights to position",
                ));
            }
            let Position { x, y } = Request::params(method, params)?;
            let label = entry.label.clone();
            let entry = ctx.windows.require_mut(&label)?;
            Ok(json!(entry.set_traffic_lights(Some((x, y)))))
        }

        // Hand the window buttons to the application, or take them back.
        // Answers with the metrics, whose `insetLeft` is then zero - the same
        // signal the platforms without any already give.
        "window.setTitleBarButtons" => {
            let Buttons { buttons } = Request::params(method, params)?;
            let label = entry.label.clone();
            let entry = ctx.windows.require_mut(&label)?;
            Ok(json!(entry.set_buttons_hidden(
                buttons == crate::config::TitleBarButtons::Hidden
            )))
        }
        "window.titleBarButtons" => Ok(json!(if entry.buttons_hidden {
            "hidden"
        } else {
            "system"
        })),

        // Back to the middle of the bar, which is where they belong unless
        // an application has a reason otherwise.
        "window.centerTrafficLights" => {
            let label = entry.label.clone();
            let entry = ctx.windows.require_mut(&label)?;
            Ok(json!(entry.set_traffic_lights(None)))
        }

        "window.openDevtools" => {
            #[cfg(feature = "devtools")]
            {
                entry.webview.open_devtools();
                return Ok(Value::Null);
            }
            #[allow(unreachable_code)]
            Err(ApiError::unsupported(
                "Devtools are not available in this build",
            ))
        }

        _ => Err(ApiError::unknown_method(method)),
    }
}

/// The window a call is about: the one it names, or the one that made it.
fn target_label(ctx: &MainCtx<'_>, params: &Value) -> String {
    serde_json::from_value::<Target>(params.clone())
        .unwrap_or_default()
        .label
        .unwrap_or_else(|| ctx.source.to_string())
}

fn create(ctx: &mut MainCtx<'_>, params: Value) -> ApiResult {
    let params: CreateParams = Request::params("window.create", params)?;
    validate_label(&params.label)?;

    if ctx.windows.get(&params.label).is_some() {
        return Err(ApiError::new(
            crate::error::code::ALREADY_EXISTS,
            format!("A window labelled `{}` is already open", params.label),
        ));
    }

    let url = windows::resolve_url(ctx.rt, params.url.as_deref());
    let proxy = ctx.proxy();

    ctx.windows
        .create(
            ctx.target,
            ctx.rt,
            proxy,
            &params.label,
            &params.window,
            &url,
        )
        .map_err(ApiError::internal)?;

    // Windows hangs the application menu off each window.
    if let Some(entry) = ctx.windows.get(&params.label) {
        ctx.chrome.attach(&entry.window);
    }

    announce(ctx, "window.created", &params.label);
    Ok(json!(params.label))
}

fn close(ctx: &mut MainCtx<'_>, params: Value) -> ApiResult {
    let label = target_label(ctx, &params);
    if ctx.windows.get(&label).is_none() {
        // Closing an already-closed window is not an error worth raising.
        return Ok(json!(false));
    }

    ctx.windows.close(&label);
    announce(ctx, "window.closed", &label);

    if ctx.windows.is_empty() && ctx.rt.config.quit_on_last_window_closed {
        *ctx.exit = true;
    }
    Ok(json!(true))
}

/// Tell every remaining window that the set of windows changed.
fn announce(ctx: &MainCtx<'_>, event: &str, label: &str) {
    ctx.windows.deliver(
        None,
        &Outgoing::Event(Event::new(event, json!({ "label": label }))),
    );
}

/// Labels end up in URLs, logs and error messages, so keep them boring.
fn validate_label(label: &str) -> Result<(), ApiError> {
    if label.is_empty() || label.len() > 64 {
        return Err(ApiError::invalid_params(
            "A window label must be between 1 and 64 characters",
        ));
    }
    if !label
        .chars()
        .all(|c| c.is_ascii_alphanumeric() || c == '-' || c == '_')
    {
        return Err(ApiError::invalid_params(format!(
            "Invalid window label `{label}`: use letters, digits, `-` and `_`"
        )));
    }
    Ok(())
}

/// What the runtime does when a window's own close button is pressed.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CloseAction {
    /// Destroy the window.
    Destroy,
    /// Hide it and tell the application, so a tray app keeps running.
    Hide,
    /// Do nothing but tell the application, which now owns the decision.
    Ignore,
}

/// Decide what a close button means for a window.
///
/// A pure function so the rule can be tested; the event loop only carries it
/// out.
pub fn close_action(behavior: CloseBehavior) -> CloseAction {
    match behavior {
        CloseBehavior::Close => CloseAction::Destroy,
        CloseBehavior::Hide => CloseAction::Hide,
        CloseBehavior::Ask => CloseAction::Ignore,
    }
}

impl CloseAction {
    /// What `window.closeRequested` reports, or `None` when the window is
    /// about to be destroyed and there is nobody left to tell.
    pub fn outcome(self) -> Option<&'static str> {
        match self {
            CloseAction::Destroy => None,
            CloseAction::Hide => Some("hidden"),
            CloseAction::Ignore => Some("ignored"),
        }
    }
}

/// Used by the event loop when the user closes a window with its own controls.
pub fn closed_by_user(
    windows: &mut crate::windows::WindowManager,
    label: &str,
    quit_when_empty: bool,
) -> bool {
    windows.close(label);
    windows.deliver(
        None,
        &Outgoing::Event(Event::new("window.closed", json!({ "label": label }))),
    );
    windows.is_empty() && quit_when_empty
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_close_button_destroys_the_window_by_default() {
        assert_eq!(close_action(CloseBehavior::Close), CloseAction::Destroy);
        // Nothing to tell: the window is going away.
        assert_eq!(CloseAction::Destroy.outcome(), None);
    }

    #[test]
    fn hide_keeps_the_webview_alive_and_says_so() {
        // This is what lets a tray application survive its window closing:
        // the webview is still there, so timers keep firing and sockets stay
        // connected.
        assert_eq!(close_action(CloseBehavior::Hide), CloseAction::Hide);
        assert_eq!(CloseAction::Hide.outcome(), Some("hidden"));
    }

    #[test]
    fn ask_hands_the_decision_to_the_application() {
        assert_eq!(close_action(CloseBehavior::Ask), CloseAction::Ignore);
        assert_eq!(CloseAction::Ignore.outcome(), Some("ignored"));
    }

    #[test]
    fn window_labels_have_to_be_boring() {
        assert!(validate_label("settings").is_ok());
        assert!(validate_label("preview-2_x").is_ok());
        assert!(validate_label("").is_err());
        assert!(validate_label("not a label").is_err());
        assert!(validate_label("../escape").is_err());
        assert!(validate_label(&"x".repeat(65)).is_err());
    }
}
